### ISSUE TITLE
Use `--deep` option to look-up sites info for given dataset (rucio container)

### ISSUE THREAD

**Issue by vkuznet** [dmwm/das2go#53]
According to discussion in this thread https://github.com/cms-sw/cmssw/issues/40889#issuecomment-1446638392 DAS provides different set of site results for two types of DAS queries:
- site file=file.root
- and, `file dataset=/a/b/c site=XYZ`

The cause of this lies in usage of `--deep` flag in URL call like this `http://cms-rucio.cern.ch/replicas/cms/RelValTTbar_14TeV/CMSSW_12_3_0_pre5-123X_mcRun4_realistic_v4_2026D88noPU-v1/GEN-SIM` (second use-case above). We need Rucio team assistance, e.g. from @ericvaandering , to provide clear REST API option for this `--deep` python flag. Once REST API call will be clarified then it can be incorporated into DAS codebase.

**Comment by ericvaandering** [comment]
Commented on the referenced issue

**Comment by vkuznet** [comment]
Added `deep` option in this commit 6e74f2ec5c8ddb5558b748c77276b7c2692f9484

**Comment by todor-ivanov** [comment]
Hi @vkuznet, 

Upon working with my dear friend ChatGPT ;) on that issue for a while and inspecting the current code, and the behaviour of both DAS and Rucio client, and reading the official Rucio documentation as well etc. ..., here is what we find: 

  The original change in 6e74f2ec5c8ddb5558b748c77276b7c2692f9484 only appended `?deep=True` to the URL generated for `rucio:file4dataset_site`. It did not reproduce the full behavior of the Rucio client.

#### Using the following parameters for our investigation:

**Dataset:**  `/TTto2L2Nu_TuneCP5_13p6TeV_powheg-pythia8/Run3Summer22EENanoAODv12-130X_mcRun3_2022_realistic_postEE_v6-v2/NANOAODSIM`

  and 

**RSE:** `T2_CH_CERN`, 

#### The current DAS code constructs:

  `http://cms-rucio.cern.ch/replicas/cms/TTto2L2Nu_TuneCP5_13p6TeV_powheg-pythia8/Run3Summer22EENanoAODv12-130X_mcRun3_2022_realistic_postEE_v6-v2/NANOAODSIM?deep=True`

#### The server rejects this request with HTTP 400:

  `ValueError: Could not parse ... into scope and name`

  The URL loses the distinction between the `/` separating the `cms` scope from the name and the leading `/` belonging to the CMS dataset name.

####  A correctly encoded top-level request:

  `https://cms-rucio.cern.ch/replicas/cms/%2FTTto2L2Nu_TuneCP5_13p6TeV_powheg-pythia8%2FRun3Summer22EENanoAODv12-130X_mcRun3_2022_realistic_postEE_v6-v2%2FNANOAODSIM/datasets?deep=True`

  returns HTTP 200 but an empty stream. This is not the sequence used by Rucio client 39.2.0 for a container.

####  The client performs:

  1. `GET /dids/cms/<encoded-name>/meta`
  2. `GET /dids/cms/?name=<encoded-name>&recursive=True`
  3. For every resolved child/block DID:
     `GET /replicas/cms/<encoded-child-name>/datasets?deep=True`

  The child requests return the expected replica information; for example, the shown child block reports `T2_CH_CERN` as `FOUND=9`, `TOTAL=9`.

###  Code audit results:

  - Rucio metadata/type detection and recursive container expansion are absent.
  - Equivalent dataset-to-block expansion through DBS already exists for `site dataset=...`.
  - Encoded per-block `/datasets?deep=True` requests already exist in `services/combined.go` and `services/rucio.go`.
  - `POST /replicas/list` plus intersection with DBS files already exists for combined `file dataset/block run site` queries.
  - That `/replicas/list` mechanism is not reachable for plain `file dataset=... site=...`.
  - The direct `file4dataset_site` unmarshaller accepts a site whenever it appears in `states`, without requiring `AVAILABLE`; this matches the historical CMSSW report where a `BEING_DELETED` replica was returned as a file at T2_CH_CERN.
  - The shared HTTP fetcher does not validate response status. The current HTTP 400 response is therefore converted into an empty result and the DAS request can still finish with status `ok`.

  Therefore this issue is not solved by just appending `deep=True` alone to the url construction from the `file4dataset*` APIs.


**Few questions follow up here:**

*  For `file dataset=... site=...`, `/datasets?deep=True` returns block/RSE replica summaries, not file records. Before implementation, we should decide whether to:

  1. reproduce the CLI sequence—type lookup, recursive child expansion, then per-child deep replica lookup plus file enumeration; or
  2. reuse the existing `POST /replicas/list` file-level mechanism already used by the combined run/site queries.

  The second path appears closer to the required DAS output, but it must be evaluated for the complete `file4dataset`, `file4dataset_site`, `file4block`, and `file4block_site` query family rather than introduced as another query-specific branch.

*  HTTP status handling should also be corrected independently so Rucio 4xx/5xx responses cannot silently become successful empty DAS results.

**Comment by todor-ivanov** [comment]
ok, 

before taking any decision to switch API calls to Rucio inside the respective DAS functionality and implementation, let give some more details on what maps to what for those two Rucio APIs: 

### The two APIs answer different questions:

|  API  |  Question answered  | Output granularity |
| :--- | :--- | :--- |
|GET /replicas/{scope}/{name}/datasets?deep=True   |  “At which RSEs is this Rucio dataset present, and how complete is it?”      |     One summary per dataset/block and RSE | 
| POST /replicas/list                              |  “Which files under these DIDs have usable replicas, and what are their PFNs?” |    One record per file |

 ### 1. `/datasets?deep=True`

  **Example:**

```
  GET /replicas/cms/%2FA%2FB%2FC%23block-id/datasets?deep=True
  Accept: application/x-json-stream
```
 Rucio calls list_dataset_replicas(scope, name, deep=True). With deep=False, it reads the collection-replica summary table. With deep=True, it recalculates the summary from the underlying available file replicas.

**Representative JSON-stream row:**
```
  {
    "scope": "cms",
    "name": "/A/B/C#block-id",
    "rse": "T2_CH_CERN",
    "rse_id": "...",
    "bytes": 9000,
    "length": 9,
    "available_bytes": 9000,
    "available_length": 9,
    "state": "AVAILABLE",
    "created_at": "...",
    "updated_at": "...",
    "accessed_at": null
  }
```

**Important properties:**

  - It returns aggregate counts, not file names.
  - length is the total number of files.
  - available_length is the number with an available replica at that RSE.
  - state is synthesized as AVAILABLE only when available_length == length.
  - It does not accept an rse_expression filter.
  - deep=True means “calculate using file-level replica rows”; it does not mean “recursively traverse arbitrary containers.”

  #### CMS container problem

  A CMS dataset DID such as: `cms:/Primary/Processed/TIER`   is normally a Rucio container whose children are block DIDs, while each block is a Rucio dataset.

Therefore:

  `GET /replicas/cms/%2FPrimary%2FProcessed%2FTIER/datasets?deep=True`

  is being applied at the wrong Rucio type/granularity. The deep implementation examines files directly belonging to the supplied dataset DID. It does not recursively descend:

 `container → block datasets → files`

  That explains why the correctly encoded top-level call in the investigation returned an empty stream.
  It works after resolving the container into blocks and calling it for each block:

  `GET /replicas/cms/%2FPrimary%2FProcessed%2FTIER%23block-id/datasets?deep=True`

  The result is then a FOUND/TOTAL-style block/RSE summary, still not a file list.

### 2. `POST /replicas/list`

  **Example:**
```
  POST /replicas/list
  Content-Type: application/json
  Accept: application/x-json-stream

  {
    "dids": [
      {
        "scope": "cms",
        "name": "/Primary/Processed/TIER"
      }
    ],
    "rse_expression": "T2_CH_CERN",
    "all_states": false,
    "resolve_archives": true,
    "resolve_parents": false
  }
```
  Rucio describes this as “list file replicas for a list of DIDs.” It accepts file, dataset, and container DIDs. For collection DIDs, its core implementation resolves contained files before retrieving their replicas.

**Representative output row:**

```
  {
    "scope": "cms",
    "name": "/store/mc/.../file.root",
    "bytes": 123456789,
    "md5": null,
    "adler32": "12ab34cd",
    "states": {
      "T2_CH_CERN": "AVAILABLE"
    },
    "rses": {
      "T2_CH_CERN": [
        "davs://eoscms.cern.ch/.../file.root"
      ]
    },
    "pfns": {
      "davs://eoscms.cern.ch/.../file.root": {
        "rse": "T2_CH_CERN",
        "rse_id": "...",
        "type": "DISK",
        "domain": "wan",
        "priority": 1,
        "volatile": false,
        "client_extract": false
      }
    }
  }
```

  **Important properties:**

  - It emits one row per file, not one row per RSE summary.
  - It supports multiple input DIDs in one request.
  - A dataset or container input is expanded into its contained files.
  - rse_expression performs the site selection server-side.
  - With all_states=false, only AVAILABLE replica states are selected.
  - With all_states=true, states such as BEING_DELETED, UNAVAILABLE, and COPYING may appear.
  - It provides file metadata, PFNs, RSEs, and replica states.
  - Both APIs return newline-delimited JSON streams rather than a conventional JSON array.

  A subtle distinction: for a collection input, list_replicas returns contained files that have matching replicas. It is not a complete collection-membership listing. Files with no replica matching the request are omitted.

## In summary:

- `/datasets?deep=True` performs server-side aggregation of file replicas and returns one completeness summary per RSE. It has no RSE filter, so DAS must select the desired RSE after receiving those summary rows.
- `POST /replicas/list` performs server-side DID expansion and RSE filtering, but returns file-level records. If DAS needs aggregate statistics, it must calculate them client-side from those records.

So:

```
/datasets?deep=True
files → server aggregation → per-RSE summaries → client RSE filtering
```

```
POST /replicas/list
DID expansion + server RSE filtering → individual file-replica records → optional client aggregation
```

 Also, `deep=True` works properly for Rucio dataset DIDs — CMS blocks, not directly for a top-level CMS container.

**Comment by todor-ivanov** [comment]
And to be complete in our summary, here follows a full map of `DAS Functionality` to `RUCIO API`:

**Legend:**
 
```
  <D> = CMS dataset/container
  <B> = CMS block / Rucio dataset
  <F> = file
  <S> = site/RSE
```

### Direct Rucio mappings:
```
  dataset4site
    DAS: dataset site=<S>
    -> GET /replicas/rse/<S>
```
```
  block4site
    DAS: block site=<S>
    -> GET /replicas/rse/<S>
```
```
  block4dataset
    DAS: block dataset=<D>
    -> GET /dids/cms/<D>/dids
    -> per block: GET /replicas/cms/<B>/datasets?deep=True
```
```
  block4dataset_site
    DAS: block dataset=<D> site=<S>
    -> GET /dids/cms/<D>/dids
    -> per block: GET /replicas/cms/<B>/datasets?deep=True
    -> DAS filters summaries by <S>
```
```
  dataset4dataset
    DAS: dataset dataset=<D>
    -> GET /dids/cms/<D>/dids
    -> per block: GET /replicas/cms/<B>/datasets?deep=True
    -> DAS aggregates block summaries
```
```
  dataset4dataset_site
    DAS: dataset dataset=<D> site=<S>
    -> GET /dids/cms/<D>/dids
    -> per block: GET /replicas/cms/<B>/datasets?deep=True
    -> DAS filters by <S> and aggregates
```
```
  block4block
    DAS: block block=<B>
    -> GET /replicas/cms/<B>/datasets?deep=True
```
```
  rules4dataset
    DAS: rules dataset=<D>
    -> GET /dids/cms/<D>/rules
```
```
  rules4block
    DAS: rules block=<B>
    -> GET /dids/cms/<B>/rules
```
```
  rules4file
    DAS: rules file=<F>
    -> GET /dids/cms/<F>/rules
```
```
  site4block
    DAS: site block=<B>
    -> GET /replicas/cms/<B>
    -> DAS extracts RSEs from file-replica records
```
```
  site4file
    DAS: site file=<F>
    -> GET /replicas/cms/<F>
    -> DAS extracts RSEs
```
```
  file4dataset
    DAS: file dataset=<D>
    -> GET /replicas/cms/<D>
```
```
  file4dataset_site
    DAS: file dataset=<D> site=<S>
    -> GET /replicas/cms/<D>?deep=True
    -> DAS filters file records by <S>
    NOTE: Rucio's direct GET handler ignores deep=True
```
```
  file4block
    DAS: file block=<B>
    -> GET /replicas/cms/<B>
```
```
  file4block_site
    DAS: file block=<B> site=<S>
    -> GET /replicas/cms/<B>
    -> DAS filters file records by <S>
```
```
  rses
    DAS: site=<S-pattern>
    -> GET /rses/
    -> DAS performs local name matching
```
```
  accounts
    DAS: no currently selectable DAS query
    -> /accounts is declared, but has no DAS-key mapping
```



  **NOTE:**  `file4dataset_site` appends `?deep=True` to `GET /replicas/<D>`, not to the `/datasets` endpoint. Rucio 39.2.0’s direct `GET /replicas/<D>` handler does not process a deep parameter, so that option is effectively _**ignored**_.

### Combined mappings that call Rucio

```
  site4dataset_pct
    DAS: site dataset=<D>
    -> DBS obtains totals and block names
    -> for each block: GET /replicas/cms/<B>/datasets?deep=True
    -> DAS aggregates site completeness
```
```
  site4dataset
    DAS: site dataset=<D>
    -> current combined implementation returns empty
    NOTE: site4dataset_pct supplies the useful result
```
```
  site4block
    DAS: site block=<B>
    -> current combined implementation returns empty
    NOTE: direct rucio:site4block supplies the useful result
```

```
  files4dataset_runs_site
    DAS: file dataset=<D> run=<R> site=<S>
    -> DBS enumerates files matching dataset/run
    -> POST /replicas/list with <D> and <S>
    -> DAS intersects both file lists
```
```
  files4block_runs_site
    DAS: file block=<B> run=<R> site=<S>
    -> DBS enumerates files matching block/run
    -> POST /replicas/list with parent <D> and <S>
    -> DAS intersects both file lists
```

**NOTE:**  The combined `site4dataset` and `site4block` functions currently return empty results.
